sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/api/FunMatcher.kt

```diff
@@ -42,21 +42,26 @@ private const val VARARG_PREFIX = "vararg ";
 class FunMatcherImpl(
     // In case of Top Level function there is no type there,
     // so you just need to specify the package
-    val qualifier: String? = null,
+//val qualifierx: String? = null,
+    val qualifiers: Set<String> = emptySet(),
```

Comment:
**As for the FunMatcher arguments,** I think you are right. I even think we should remove them as arguments entirely and rely on the typesafe builder. The only argument to keep them is to avoid refactoring. Reasons:

1. We do now have two different approaches to how we can achieve the same thing with the API: using parameters or using the typesafe builder. Good API design should try to avoid that, as to avoid disambiguity in how the user should use the API, and also, to avoid redundancy.
2. Avoid boilerplate code in the function matcher. We have defined the whole set of parameters (`qualifier`, `name`, `arguments`, ...) in three different locations: `FunMatcherImpl`, `FunMatcherBuilderContext` and `FunMatcher` functions.

In the `FieldMatcher`, I skipped such arguments entirely and you only have instance fields (declared once) that you can access from within the typesafe builder.

**As for the matching logic:** I do not see any chance how to simplify that. There is no method `withQualifiers` at the moment. If we want to introduce one, we still need a `qualifiers: set<String>` and a change in the matching logic. The only difference is that `qualifiers` is not a constructor argument anymore.

Reply:
P.S.: since all these properties are copied from the `FunMatcherBuilderContext` to the `FunMatcherBuilderImpl`, it would not make sense to remove `qualifiers` from the `FunMatcherImpl` constructor. We could move it only from the `FunMatcherBuilderContext` constructor only and make it a regular instance field there, and remove it from the `FunMatcher` function as an argument. 

However, `qualifiers` would be the only FunMatcher property that is treated differently then, and I don't know if that makes sense. 

My suggestion would therefore be: no special treat for `qualifiers` (i.e., not fixed), but maybe we should refactor the `FunMatcher` later to enforce the use of the typesafe builder. Wdyt?

Reply:
As discussed: removed lists and sets from the parameter list and refactored the code to use typesafe builder. 

I did this even for `names` and `arguments` and refactored the respective checks. 

Fixed.